<a href="https://colab.research.google.com/github/JosenildoJunior/StatPyDataScience/blob/main/Correla%C3%A7%C3%A3o_e_Regress%C3%A3o.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Estatística com Python: Um Guia para Estudos e Solução de Problemas**

Breve resumo:

Correlação e regressão são duas técnicas estatísticas usadas para medir e modelar a relação entre duas ou mais variáveis.

Correlação

- A correlação mede a força e a direção da relação entre duas variáveis. O coeficiente de correlação é um número entre -1 e 1. Um coeficiente de correlação próximo a 1 indica uma forte relação positiva entre as duas variáveis, um coeficiente de correlação próximo a -1 indica uma forte relação negativa entre as duas variáveis, e um coeficiente de correlação próximo a 0 indica que não há relação entre as duas variáveis.

Regressão

- A regressão é usada para modelar a relação entre uma variável dependente e uma ou mais variáveis independentes. A equação de regressão é uma equação que descreve a relação entre as variáveis.

Quando utilizar:

- Correlação: Sempre que se deseja medir a força e a direção da relação entre duas variáveis.

- Regressão: Sempre que se deseja modelar a relação entre uma variável dependente e uma ou mais variáveis independentes.
Exemplos:

- Correlação: Um pesquisador pode usar a correlação para medir a relação entre o nível de escolaridade e o salário.

- Regressão: Um economista pode usar a regressão para modelar a relação entre o preço da gasolina e a demanda por gasolina.

# **Correlação**

## **Importações**

Para exemplificação dos cálculos de correlação será necessário a importação de algumas bibliotecas.

In [ ]:
# Álgebra linear
import numpy as np

# Manipualçao de dados
import pandas as pd

# Visualização de dados
import seaborn as sns

# Funções matemáticas
import math

## **Base de dados**

Com as devidas importações realizadas, vamos gerar a base de dados.

In [ ]:
# Gerando os dados
tamanho = np.array([30, 39, 49, 60])
preco = np.array([57000, 69000, 77000, 90000])

Agora, vamos transformar os dados em um DataFrame.

In [ ]:
# Convertendo em um data frame
df = pd.DataFrame({'Tamanho': tamanho, 'Preco': preco})

# Observando os valores
df

tamanho  preco
0       30  57000
1       39  69000
2       49  77000
3       60  90000

## **Cálculos**

Para a execução dos cálculos, vamos utilizar as bibliotecas que importamos anteriormente.

### **Covariância**

A covariância é uma medida da relação linear entre duas variáveis. Ela é calculada como a média dos produtos das diferenças entre os valores das variáveis.

A covariância pode ser positiva, negativa ou nula. Uma covariância positiva indica que as duas variáveis tendem a se mover na mesma direção, enquanto uma covariância negativa indica que as duas variáveis tendem a se mover em direções opostas. Uma covariância nula indica que não há relação linear entre as duas variáveis.

In [ ]:
# Calculando a covariância
df.cov()

tamanho        preco
tamanho     167.0     178500.0
preco    178500.0  192250000.0

Podemos observar uma covariância positiva entre o tamanho e o preço de 178.500. Logo, é correto afirmar que o valor da covariância é maior que 1. Podemos afirmar que as duas variáveis tendem a se mover na mesma direção.

### **Correlação**

A matriz de correlação é uma medida normalizada que varia de -1 a 1 e indica a força e a direção da relação linear entre duas variáveis. Um valor próximo de 1 indica uma correlação positiva perfeita, -1 indica uma correlação negativa perfeita, e 0 indica falta de correlação linear. Vamos observar como classificar a força dessa correlação:

Força da Correlação:

- 0.00 a 0.19 (ou -0.19): Correlação muito fraca.
- 0.20 a 0.39 (ou -0.39): Correlação fraca.
- 0.40 a 0.59 (ou -0.59): Correlação moderada.
- 0.60 a 0.79 (ou -0.79): Correlação forte.
- 0.80 a 1.00 (ou -1.00): Correlação muito forte.

In [ ]:
# Calculando a matriz de correlação
df.corr()

tamanho     preco
tamanho  1.000000  0.996201
preco    0.996201  1.000000

Obtivemos uma correlação positiva entre o tamanho e o preço de 0,99. Isso indica que ambas as variáveis se movem na mesma direção. Podemos dizer também que as duas variáveis estão correlacionadas quase perfeitamente.

## **Exercício**

 Carregue a base de dados das casas (house-prices.csv) e realize os seguintes passos:

- Gere as correlações entre todos os atributos.
- Quais atributos estão mais relacionados com o preço das casas?

Para a realização desse exercício, será necessário, primeiramente, a importação das bibliotecas que serão utilizadas.

In [ ]:
# Álgebra linear
import numpy as np

# Manipualçao de dados
import pandas as pd

# Visualização de dados
import seaborn as sns

Com todas essas bibliotecas importadas, podemos seguir adiante para a importação dos dados.

In [ ]:
# Acesso ao drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Importando o dataset
df = pd.read_csv("/content/drive/MyDrive/Estatística para ciência de dados/house_prices.csv")

# Observando os registros
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Essa base possui 21 variáveis, entre elas podem haver variáveis não numéricas. Vamos verificar essa possibilidade.

In [ ]:
# Observando os tipos de dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

Algumas observações:

- É possível observar que a variável date não é do tipo numérico. Portanto, por enquanto, vamos retirar essa variável do data frame.

- A variável id provavelmente não apresentará nenhum tipo de correlação, pois se trata de um atributo de identificação. Logo, podemos excluí-la da base também.

- Aparentemente, as variáveis sqft_living e sqft_lot estão mais desatualizadas do que as variáveis sqft_living15 e sqft_lot15. Sendo assim, vou excluir também essas que estão mais antigas.

In [ ]:
# Realizando a exclusão da coluna
df = df.drop(['date', 'id', 'sqft_living', 'sqft_lot'], axis = 1)

# Observando os dados
df.head()

price  bedrooms  bathrooms  floors  waterfront  view  condition  grade  \
0  221900.0         3       1.00     1.0           0     0          3      7   
1  538000.0         3       2.25     2.0           0     0          3      7   
2  180000.0         2       1.00     1.0           0     0          3      6   
3  604000.0         4       3.00     1.0           0     0          5      7   
4  510000.0         3       2.00     1.0           0     0          3      8   

   sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0        1180              0      1955             0    98178  47.5112   
1        2170            400      1951          1991    98125  47.7210   
2         770              0      1933             0    98028  47.7379   
3        1050            910      1965             0    98136  47.5208   
4        1680              0      1987             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  
0 -122.257           1340        5650  
1 -122.319           1690        7639  
2 -122.233           2720        8062  
3 -122.393           1360        5000  
4 -122.045           1800        7503

Com as devidas remoções realizadas, podemos seguir adiante para o cálculo propriamente dito.

In [ ]:
# Calculando as correlações
correlacoes = df.corr()

# Exibindo os resultados
correlacoes

price  bedrooms  bathrooms    floors  waterfront      view  \
price          1.000000  0.308350   0.525138  0.256794    0.266369  0.397293   
bedrooms       0.308350  1.000000   0.515884  0.175429   -0.006582  0.079532   
bathrooms      0.525138  0.515884   1.000000  0.500653    0.063744  0.187737   
floors         0.256794  0.175429   0.500653  1.000000    0.023698  0.029444   
waterfront     0.266369 -0.006582   0.063744  0.023698    1.000000  0.401857   
view           0.397293  0.079532   0.187737  0.029444    0.401857  1.000000   
condition      0.036362  0.028472  -0.124982 -0.263768    0.016653  0.045990   
grade          0.667434  0.356967   0.664983  0.458183    0.082775  0.251321   
sqft_above     0.605567  0.477600   0.685342  0.523885    0.072075  0.167649   
sqft_basement  0.323816  0.303093   0.283770 -0.245705    0.080588  0.276947   
yr_built       0.054012  0.154178   0.506019  0.489319   -0.026161 -0.053440   
yr_renovated   0.126434  0.018841   0.050739  0.006338    0.092885  0.103917   
zipcode       -0.053203 -0.152668  -0.203866 -0.059121    0.030285  0.084827   
lat            0.307003 -0.008931   0.024573  0.049614   -0.014274  0.006157   
long           0.021626  0.129473   0.223042  0.125419   -0.041910 -0.078400   
sqft_living15  0.585379  0.391638   0.568634  0.279885    0.086463  0.280439   
sqft_lot15     0.082447  0.029244   0.087175 -0.011269    0.030703  0.072575   

               condition     grade  sqft_above  sqft_basement  yr_built  \
price           0.036362  0.667434    0.605567       0.323816  0.054012   
bedrooms        0.028472  0.356967    0.477600       0.303093  0.154178   
bathrooms      -0.124982  0.664983    0.685342       0.283770  0.506019   
floors         -0.263768  0.458183    0.523885      -0.245705  0.489319   
waterfront      0.016653  0.082775    0.072075       0.080588 -0.026161   
view            0.045990  0.251321    0.167649       0.276947 -0.053440   
condition       1.000000 -0.144674   -0.158214       0.174105 -0.361417   
grade          -0.144674  1.000000    0.755923       0.168392  0.446963   
sqft_above     -0.158214  0.755923    1.000000      -0.051943  0.423898   
sqft_basement   0.174105  0.168392   -0.051943       1.000000 -0.133124   
yr_built       -0.361417  0.446963    0.423898      -0.133124  1.000000   
yr_renovated   -0.060618  0.014414    0.023285       0.071323 -0.224874   
zipcode         0.003026 -0.184862   -0.261190       0.074845 -0.346869   
lat            -0.014941  0.114084   -0.000816       0.110538 -0.148122   
long           -0.106500  0.198372    0.343803      -0.144765  0.409356   
sqft_living15  -0.092824  0.713202    0.731870       0.200355  0.326229   
sqft_lot15     -0.003406  0.119248    0.194050       0.017276  0.070958   

               yr_renovated   zipcode       lat      long  sqft_living15  \
price              0.126434 -0.053203  0.307003  0.021626       0.585379   
bedrooms           0.018841 -0.152668 -0.008931  0.129473       0.391638   
bathrooms          0.050739 -0.203866  0.024573  0.223042       0.568634   
floors             0.006338 -0.059121  0.049614  0.125419       0.279885   
waterfront         0.092885  0.030285 -0.014274 -0.041910       0.086463   
view               0.103917  0.084827  0.006157 -0.078400       0.280439   
condition         -0.060618  0.003026 -0.014941 -0.106500      -0.092824   
grade              0.014414 -0.184862  0.114084  0.198372       0.713202   
sqft_above         0.023285 -0.261190 -0.000816  0.343803       0.731870   
sqft_basement      0.071323  0.074845  0.110538 -0.144765       0.200355   
yr_built          -0.224874 -0.346869 -0.148122  0.409356       0.326229   
yr_renovated       1.000000  0.064357  0.029398 -0.068372      -0.002673   
zipcode            0.064357  1.000000  0.267048 -0.564072      -0.279033   
lat                0.029398  0.267048  1.000000 -0.135512       0.048858   
long              -0.068372 -0.564072 -0.135512  1.000000       0.334605   
sqft_livin

Agora, podemos observar as correlações entre todas as variáveis. Dessa forma, podemos observar alguns detalhes interessantes, como por exemplo:

- Bathrroms e bedrooms possuem uma correlação positiva moderada de 0,51.
- Floors e Bathrooms também apresentam uma correlação positiva moderada de 0,50.
- A variável sqft_living e banheiros apresentam uma correlação positiva moderada de 0,56.
- A variável grade e bathrooms apresentam uma correlação positiva forte de 0,66.
- A variável grade e price apresentam uma correlação positiva forte de 0,66.

Existem outras correlações que podemos apontar, mas como nosso objetivo é observar as correlações entre a variável 'price' e as outras variáveis, vamos observar apenas as correlações com a variável 'price'.

In [ ]:
# Visualizando as correlações
correlacoes.head(1)

price  bedrooms  bathrooms    floors  waterfront      view  condition  \
price    1.0   0.30835   0.525138  0.256794    0.266369  0.397293   0.036362   

          grade  sqft_above  sqft_basement  yr_built  yr_renovated   zipcode  \
price  0.667434    0.605567       0.323816  0.054012      0.126434 -0.053203   

            lat      long  sqft_living15  sqft_lot15  
price  0.307003  0.021626       0.585379    0.082447

Podemos observar que as maiores correlações com o preço são:

- grade apresenta uma correlação positiva forte de 0,66.
- sqft_above apresenta uma correlação positiva forte de 0,60.
- sqft_living apresenta uma correlação positiva moderada de 0,58.
- bathrooms apresenta uma correlação positiva moderada de 0,52.

Essas são as quatro variáveis que mais apresentam correlações com a variável 'price'. É importante frisar que todas essas correlações são positivas.

# **Regressão**

Algumas observasões sobre a regressão:

- Espera-se que a correlação seja de moderada a forte para obter um bom modelo (positiva ou negativa)
- Coeficiente de determinação (R2): > 0.7 é um bom valor. Entre 0 e 0.3 é ruim. Entre esses valores é interessante fazer testes
- Regressão linear: existir lineariedade

Vamos realizar as importações necessárias para a aplicação do modelo de regressão propriamente dito.

In [ ]:
# Álgebra linear
import numpy as np

# Manipualçao de dados
import pandas as pd

# Visualização de dados
import seaborn as sns
import matplotlib.pyplot as plt

# Funções matemáticas
import math

Com todas essas bibliotecas importadas, podemos seguir adiante para a importação dos dados.

In [ ]:
# Acesso ao drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Importando o dataset
df = pd.read_csv("/content/drive/MyDrive/Estatística para ciência de dados/house_prices.csv")

# Observando os registros
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## **Reressão Linear simples**

Excluindo as variáveis que não serão utilizadas.

In [ ]:
# Realizando a exclusão da coluna
df = df.drop(['date', 'id', 'sqft_living', 'sqft_lot'], axis = 1)

# Observando os dados
df.head()

price  bedrooms  bathrooms  floors  waterfront  view  condition  grade  \
0  221900.0         3       1.00     1.0           0     0          3      7   
1  538000.0         3       2.25     2.0           0     0          3      7   
2  180000.0         2       1.00     1.0           0     0          3      6   
3  604000.0         4       3.00     1.0           0     0          5      7   
4  510000.0         3       2.00     1.0           0     0          3      8   

   sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0        1180              0      1955             0    98178  47.5112   
1        2170            400      1951          1991    98125  47.7210   
2         770              0      1933             0    98028  47.7379   
3        1050            910      1965             0    98136  47.5208   
4        1680              0      1987             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  
0 -122.257           1340        5650  
1 -122.319           1690        7639  
2 -122.233           2720        8062  
3 -122.393           1360        5000  
4 -122.045           1800        7503

Como vimos anteriormente, essas são as quatro variáveis que apresentaram as maiores correlações:

- Grade apresenta uma correlação positiva forte de 0,66.
- Sqft_above apresenta uma correlação positiva forte de 0,60.
- Sqft_living apresenta uma correlação positiva moderada de 0,58.
- Bathrooms apresenta uma correlação positiva moderada de 0,52.

Dito isso, para a criação do nosso modelo, vamos utilizar a variável grade, que apresentou a maior correlação de todas.

Vamos separar os dados.

In [ ]:
# Definindo o valor de X
X = df['grade'].values

# Observando as dimensões
X.shape

(21613,)

Podemos observar que o nosso X é um Array unidimensional, onde tem apenas 21613 linhas. O modelo de regressão não vai aceitar esse formato, então precisaremos ajustar as dimensões dessa variável.

In [ ]:
# Ajustando a dimensão dos dados
X = X.reshape(-1, 1)

# Observando o resultado
X.shape

(21613, 1)

Podemos observar que agora temos o mesmo número de linhas, porém temos também uma coluna. Agora sim estamos com o formato que o algoritmo vai aceitar.

Definindo o valor de y.

In [ ]:
# Definindo o y
y = df['price'].values

Nesta variável não é necessário realizar a mesma transformação que utilizamos na variável X.

Nesse momento, torna-se necessário realizar a divisão entre dados de treino e dados de teste.

In [ ]:
# Bilioteca para realizção de amostras estratificadas
from sklearn.model_selection import train_test_split

# Separando os dados
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size = 0.2, random_state = 1)

Realizando a importação do modelo.

In [ ]:
# Importando o modelo
from sklearn.linear_model import LinearRegression

Realizando a importação das métricas.

In [ ]:
# Importando as métricas de validação
from sklearn.metrics import mean_absolute_error, mean_squared_error

Importamos as seguintes funções:

- A função mean_absolute_error() calcula o erro médio absoluto entre as previsões de um modelo e os valores reais. O erro médio absoluto é uma medida da distância média entre os valores previstos e os valores reais.

- A função mean_squared_error() calcula o erro médio quadrático entre as previsões de um modelo e os valores reais. O erro médio quadrático é uma medida da distância quadrada média entre os valores previstos e os valores reais.

Criando uma instância do modelo.

In [ ]:
# Criando o modelo
regressor = LinearRegression()

Treinando o modelo com os dados de treinamento.

In [ ]:
# Treinando o modelo
regressor.fit(X_treino, y_treino)

LinearRegression()

Utilizando o modelo treinado para fazer previsões nos dados de teste.

In [ ]:
# Realizando as predições
previsoes = regressor.predict(X_teste)

Analisando o desempenho do modelo.

Calculando o coeficiente de determinação R² para os dados de teste.

In [ ]:
# Observando o r2
regressor.score(X_teste, y_teste)

0.4328786526974916

Obtivemos um coeficiente de determinação R² de 0,43, o que significa que o modelo consegue explicar em média 43% da variância dos dados de teste.

In [ ]:
# Calcula o erro médio absoluto
mae = mean_absolute_error(y_teste, previsoes)
print(f"O erro médio absoluto é de: {mae} unidades")

# Calcula o erro médio quadrático
mse = mean_squared_error(y_teste, previsoes)
print(f"O erro médio quadrático é de: {mse} unidades")

O erro médio absoluto é de: 177341.25404777075 unidades
O erro médio quadrático é de: 97885695257.9813 unidades


- O nosso erro médio absoluto indica que, em média, nossa previsão erra cerca de 177 mil unidades. No nosso caso, por se tratar de preço, vamos considerar que seja em dólares. Logo, nosso modelo erra o valor da casa em média cerca de 177 mil dólares, seja para mais ou para menos.

- Já o erro médio quadrático é um pouco difícil de interpretar pelo seu valor atual. Porém, vamos extrair a raiz quadrada do mesmo para observá-lo melhor, calculando assim o RMSE.

A Root Mean Squared Error (RMSE) é uma métrica de avaliação de desempenho de modelos de previsão que mede a distância média entre os valores reais e as previsões.

In [ ]:
# Calculando o rmse
rmse = math.sqrt(mse)

print(f"O RMSE é de: {rmse} unidades")

O RMSE é de: 312866.8970312796 unidades


Com essa métrica, podemos afirmar que a diferença média entre os preços reais e os preços preditos é de 312 mil.

## **Regressão Linear múltipla**

Na regressão linear multipla, mais de uma variável independente é usada para prever a variável dependente. Por exemplo, podemos usar a regressão linear multipla para prever o preço de uma casa com base na sua área, no número de quartos e no número de banheiros.

Separando os dados

In [ ]:
# Definindo o valor de X
X = df[['grade', 'sqft_above', 'sqft_living15', 'bathrooms']].values

Definindo o valor de y.

In [ ]:
# Definindo o y
y = df['price'].values

Nesse momento, torna-se necessário realizar a divisão entre dados de treino e dados de teste.

In [ ]:
# Bilioteca para realizção de amostras estratificadas
from sklearn.model_selection import train_test_split

# Separando os dados
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size = 0.2, random_state = 1)

Criando uma instância do modelo.

In [ ]:
# Criando o modelo
regressor = LinearRegression()

Treinando o modelo com os dados de treinamento.

In [ ]:
# Treinando o modelo
regressor.fit(X_treino, y_treino)

LinearRegression()

Utilizando o modelo treinado para fazer previsões nos dados de teste.

In [ ]:
# Realizando as predições
previsoes = regressor.predict(X_teste)

Analisando o desempenho do modelo.

Calculando o coeficiente de determinação R² para os dados de teste.

In [ ]:
# Observando o r2
regressor.score(X_teste, y_teste)

0.475898660606238

Obtivemos um coeficiente de determinação R² de 0,47, o que significa que o modelo consegue explicar, em média, 47% da variância dos dados de teste. Podemos notar uma melhora de 4% em relação à regressão linear simples, já que o seu R² foi de 0,43.

In [ ]:
# Calcula o erro médio absoluto
mae = mean_absolute_error(y_teste, previsoes)
print(f"O erro médio absoluto é de: {mae} unidades")

# Calcula o erro médio quadrático
mse = mean_squared_error(y_teste, previsoes)
print(f"O erro médio quadrático é de: {mse} unidades")

O erro médio absoluto é de: 173501.72172665017 unidades
O erro médio quadrático é de: 90460400117.56529 unidades


- O nosso erro médio absoluto indica que, em média, nossa previsão erra cerca de 173 mil unidades. No nosso caso, por se tratar de preço, vamos considerar que seja em dólares. Logo, nosso modelo erra o valor da casa em média cerca de 173 mil dólares, seja para mais ou para menos.

Podemos notar uma sutil melhora aqui também, mas vamos analisar o RMSE para entendermos melhor.

In [ ]:
# Calculando o rmse
rmse = math.sqrt(mse)

print(f"O RMSE é de: {rmse} unidades")

O RMSE é de: 300766.3546967401 unidades


Com essa métrica, podemos afirmar que o erro médio entre os preços reais e os preços preditos é de 300 mil, uma melhora de 12 mil em relação ao modelo anterior.

## **Exercício**

O objetivo deste exercício é criar um modelo de regressão linear múltipla utilizando todos os atributos da base de dados.

**Carregando os dados**

Realizando as importações necessárias para a aplicação do modelo de regressão propriamente dito.

In [1]:
# Álgebra linear
import numpy as np

# Manipualçao de dados
import pandas as pd

# Visualização de dados
import seaborn as sns
import matplotlib.pyplot as plt

# Funções matemáticas
import math

Com todas essas bibliotecas importadas, podemos seguir adiante para a importação dos dados.

In [2]:
# Acesso ao drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
# Importando o dataset
df = pd.read_csv("/content/drive/MyDrive/Estatística para ciência de dados/house_prices.csv")

# Observando os registros
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

**Pré-Processamento**

Nessa etapa, vamos separar os dados nas variáveis preditoras e variáveis resposta.

In [6]:
# Separando os dados em X e Y

# Variáveis preditoras
X = df.drop(['date', 'id', 'sqft_living', 'sqft_lot', 'price'], axis = 1)

# Variável resposta
y = df['price'].values

**Machine Learning**

Realizando as importações para a divisão dos dados em treino e teste.

In [7]:
# Bilioteca para realizção de amostras estratificadas
from sklearn.model_selection import train_test_split

# Separando os dados
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size = 0.2, random_state = 4)

Importando o modelo

In [8]:
# Importando o modelo
from sklearn.linear_model import LinearRegression

Realizando a importação das métricas.

In [9]:
# Importando as métricas de validação
from sklearn.metrics import mean_absolute_error, mean_squared_error

Criando uma instância do modelo.

In [10]:
# Criando o modelo
regressor = LinearRegression()

Treinando e fazendo previsões com o modelo

In [11]:
# Treinando o modelo
regressor.fit(X_treino, y_treino)

# Realizando as predições
previsoes = regressor.predict(X_teste)

Avaliando o desempenho do modelo

Calculando o coeficiente de determinação R² para os dados de teste.

In [12]:
# Observando o r2
regressor.score(X_teste, y_teste)

0.6970140146588895

Obtivemos um R² de 0,69 nos dados de teste. Isso implica que podemos explicar 69% da variância dos dados com esse modelo. É notável também a grande melhoria se comparado com o modelo utilizando apenas 4 elementos. Aparentemente, esse modelo vai apresentar uma performance bem superior aos demais criados anteriormente. Dito isso, vamos observar outras métricas.

In [13]:
# Calcula o erro médio absoluto
mae = mean_absolute_error(y_teste, previsoes)
print(f"O erro médio absoluto é de: {mae} unidades")

# Calcula o erro médio quadrático
mse = mean_squared_error(y_teste, previsoes)
print(f"O erro médio quadrático é de: {mse} unidades")

O erro médio absoluto é de: 126383.57369183468 unidades
O erro médio quadrático é de: 39187972609.973785 unidades


- Obtivemos um erro médio absoluto (MAE) de 126 mil. Isso indica que, em média, nossas previsões podem variar cerca de 126 mil unidades, para mais ou para menos.

Podemos notar uma melhora bem promissora aqui. No entanto, vamos analisar o RMSE para entendermos melhor a disparidade entre os modelos.

In [14]:
# Calculando o rmse
rmse = math.sqrt(mse)

print(f"O RMSE é de: {rmse} unidades")

O RMSE é de: 197959.52265545042 unidades


Com essa métrica, podemos afirmar que o erro médio absoluto (MAE) entre os preços reais e os preços preditos é de 197 mil. Isso representa uma melhora de 103 mil em relação ao modelo anterior. Portanto, podemos afirmar que esse modelo apresentou um desempenho muito superior aos outros modelos criados anteriormente.

Será que é possível melhorarmos o desempenho desse modelo aplicando uma seleção de atributos? Vamos ver mais a fundo.

### **Seleção de atributos**

Primeiro, vamos importar a biblioteca para seleção de atributos.

In [33]:
# Importando os módulos
from sklearn.feature_selection import SelectFdr, f_regression

Criando o objeto seletor de atributos e definindo o valor de alpha.

In [34]:
# Criando a instância
selecao = SelectFdr(f_regression, alpha = 0.0)

Agora, podemos aplicar a seleção de atributos aos dados.

In [35]:
# Aplicando a seleção
X_novo = selecao.fit_transform(X, y)

Nesse momento, vamos criar um modelo e treiná-lo com as variáveis selecionadas. Em seguida, vamos separar a base entre treino e teste.

In [36]:
# Realizando a divisão entre os dados de treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X_novo, y, test_size = 0.2, random_state = 1)

Criando a instância do modelo.

In [37]:
# Criando o modelo
regressor_novo = LinearRegression()

Realizando o treinamento e as predições do modelo.

In [38]:
# Treinando o modelo
regressor_novo.fit(X_treino, y_treino)

# Realizando as predições
predicoes = regressor_novo.predict(X_teste)

Analisando o desempenho do modelo

In [39]:
# Observando o r2
regressor_novo.score(X_teste, y_teste)

0.6444206541785331

Nesse caso, podemos observar um R² de 0,64 que é próximo ao modelo anterior, que até o momento é o nosso melhor modelo. No entanto, vamos analisar esse aqui com mais calma.

In [40]:
# Calcula o erro médio absoluto
mae = mean_absolute_error(y_teste, predicoes)
print(f"O erro médio absoluto é de: {mae} unidades")

# Calcula o erro médio quadrático
mse = mean_squared_error(y_teste, predicoes)
print(f"O erro médio quadrático é de: {mse} unidades")

O erro médio absoluto é de: 137268.96012702258 unidades
O erro médio quadrático é de: 61373340380.61963 unidades


Podemos observar que, nesse caso, o erro médio absoluto (MAE) foi um pouco maior, cerca de 10 mil unidades. Agora, vamos observar o RMSE.

In [41]:
# Calculando o rmse
rmse = math.sqrt(mse)

print(f"O RMSE é de: {rmse} unidades")

O RMSE é de: 247736.43329276305 unidades


Com essa métrica, podemos afirmar que, nesse caso, o erro médio entre os preços reais e os preços previstos é de 247 mil unidades. Portanto, esse modelo realmente apresentou um desempenho pior que o modelo anterior. No entanto, em outros casos, talvez compense utilizar essa técnica de seleção de atributos.